### ISSUE TITLE
[TT backend] Experimental compile fails on second invocation under SPMD for sample_from_logits

### ISSUE THREAD

**Issue by kmabeeTT** [tenstorrent/tt-xla#3672]
### Problem

The experimental compile path (`_call_experimental_compile` using `torch.export` + `extract_compiled_graph`, enabled by default in #2996) fails when a compiled function is invoked a second time with different data under SPMD. In standalone repros this manifests as `Error code: 13`. In the vLLM plugin context it silently returns the warmup result.

This was discovered during #3589 when enabling `torch.compile` on `sample_from_logits` under SPMD/TP. The legacy compile path (`tt_legacy_compile=True`) does not have this issue.

### Reproduction

Under SPMD (`xr.use_spmd()`), compiling a pure function (no model parameters) with the experimental path:
```python
compiled_fn = torch.compile(fn, backend="tt", fullgraph=True, dynamic=False)
compiled_fn(torch.zeros(1, 32000).to(device))  # warmup — works
compiled_fn(torch.randn(1, 32000).to(device))  # Error code: 13
```

The HLO graph from the first call has 1 input and 1 output (confirmed via `XLA_DYNAMO_DEBUG=1`) — the input is correctly treated as a graph input, not constant-folded. But re-executing the cached graph with new data fails.

Does not reproduce:
- Without SPMD
- With legacy compile (`options={"tt_legacy_compile": True}`)
- When the function accesses model sub-modules (e.g. `gather_logprobs` works fine)

### Current workaround

`sample_from_logits` uses legacy compile via decorator option:
```python
@torch.compile(backend="tt", fullgraph=True, dynamic=False,
               options={"tt_legacy_compile": True})
def sample_from_logits(self, logits, sampling_metadata):
```

### Notes

- Debug scripts on branch `kmabee/issue_3589_spmd_debug`: `debug_constant_fold*.py`
- `gather_logprobs` uses the same warmup pattern but works — possibly because it references `self.sampler` sub-modules which changes the graph structure
- The `optimized_mod` in `dynamo_bridge.py` has SPMD-specific input sharding check logic (lines 589-604) that may be involved
